### ISSUE TITLE
Going to open a door with the full party make it drop fps considerably 

### ISSUE THREAD

**Issue by rsantellan** [gemrb/gemrb#1692]
#### Bug description
On the bridge district (Where I notice) trying to open a door brings down the fps. It's only when the full party is selected.
I believe this is related to #1441
#### Steps to reproduce
1. Go to Bridge District
2. Click on a door with the full party
3. Move to the other side and click again

#### Expected behavior
It not should drop the fps like that

#### Screenshots

https://user-images.githubusercontent.com/1820598/188778821-070c1bb5-49ea-4d2c-a260-2922323938a3.mp4

GemRB version (check as many as you know apply)
- [X] master as of this issue
- [x] 0.9.1
- [x] 0.9.0
- [x] 0.8.8

Video Driver (check as many as you know apply)
- [ ] SDL1.2
- [ ] SDL2 built with `USE_OPENGL_BACKEND`
- [X] SDL2 without `USE_OPENGL_BACKEND`

**Comment by lynxlynxlynx** [comment]
Probably _just_ the general performance problem / trade-off, unless you can demonstrate this worked better in 0.8.8 or some later version.

We'll probably need to add new threads to speed the search (eg. the direction loop in FindPath looks parallelizable), but that wouldn't help on all hardware. And since pathing is critical, we can't just wait until it's done, doing other stuff like processing mouse movement in the meanwhile.

**Comment by rsantellan** [comment]
@lynxlynxlynx I tried on all the tags beginning on the v0.8.8 and in all of them happen the same. Depending where you are the impact is more notable

**Comment by bradallred** [comment]
the profiler shows we are spending quite a bit of time in `Actor::NewPath` and half of that time is spent in `Map::NormalizeDeltas`. Not sure how to improve that.

**Comment by bradallred** [comment]
Things still hitch, but those optimizations gave me ~10% improvement (hitches are 10% shorter). We could squeeze another ~14% if we could eliminate ConvertCoordToTile from the pathfinding by having a flavor of `GetBlocked` that takes a tile coord. we'd probably have to cache the tile coord for each actor pos too.

Any better fix is going to have to actually change some logic (or offloading to another thread as mentioned)

**Comment by bradallred** [comment]
I'm not sure it is related, but this looks like a bug to me:
https://github.com/gemrb/gemrb/blob/ab71947f35bdea7e4c862bace1b9779ac6847a58/gemrb/core/Map.cpp#L2575-L2594

the `Point` passed into `GetBlockedInRadius` is already a "navmap" coordinate, but these multiplies are treating it like we are converting it from tile space.

https://github.com/gemrb/gemrb/blob/ab71947f35bdea7e4c862bace1b9779ac6847a58/gemrb/core/Map.cpp#L394-L397

Am I missing something?

**Comment by lynxlynxlynx** [comment]
Which multiplies? Anyway, either we have an issue open (or it was just chat with kmfrick) about subclassing a BasePoint, so we can have Point always be a NavMapPoint and then a separate class be the TileMapPoint to be able to canonically avoid double conversions etc.

**Comment by bradallred** [comment]
the `* 16` and `* 12`:
```
PathMapFlags retBotRight = GetBlocked(Point(p.x + i * 16, p.y + j * 12)); 
PathMapFlags retTopRight = GetBlocked(Point(p.x + i * 16, p.y - j * 12)); 
PathMapFlags retBotLeft = GetBlocked(Point(p.x - i * 16, p.y + j * 12)); 
PathMapFlags retTopLeft = GetBlocked(Point(p.x - i * 16, p.y - j * 12));
```

Yeah, distinct types would be nice to avoid bugs and for clarity

**Comment by lynxlynxlynx** [comment]
I don't see a problem there, it's iterating over the vertices of what would be the tilemap, but still in normal space.

**Comment by bradallred** [comment]
Ah, so the multiples are converting `i` into normal space to add to a normal space coord. gotcha.

**Comment by bradallred** [comment]
ok, things seem much better to me. We can go a bit further without changing the algorithm in a couple of areas too.

**Comment by rsantellan** [comment]
> ok, things seem much better to me. We can go a bit further without changing the algorithm in a couple of areas too.

Is a lot better for me, the only place is when is a narrow place that drops to <3fps.

**Comment by bradallred** [comment]
It looks like our "bumping" implementation forces the bumped actor to recalculate a path. I imagine this happens quite a bit in narrow corridors.

I may be being naïve, but it seems like we could instead find the bumped position, then calculate a path from that position to the current position in the path and then prepend that path segment, which would be _much_ better than completely recalculating a path.

**Comment by lynxlynxlynx** [comment]
Hard to say how much that would help or if it'd look good, as in tight spaces the new segment might cause bumping again or repathing.

**Comment by bradallred** [comment]
Repeated bumping and complete repathing is what appears to be happening currently. The difference is calculating a few nodes of a a new path and prepending it instead of completely recalculating the entire path as is happening now. The problem is exacerbated greatly if the whole party is pathing though a corridor with the destination being far away.

I dont know anything about how bumping is implemented, but I would hope it is at least deterministic. It's probably too much to hope that it is physics approximate such that collisions between actors behave like like products of force vectors tho :)

**Comment by MarcelHB** [comment]
Yes, the more `XYZ re-pathing ignoring actors` in the log appears, the longer to wait. Bumping-check is implemented by this: https://github.com/gemrb/gemrb/blob/71bb2f80ecbd4ce0604672b5078b7be8b2569ea8/gemrb/core/PathFinder.cpp#L374-L375

Hm, maybe we can exempt party members from this check? Especially when they gathered in front of a door already that isn't blocked by anyone else (Rylock comes to my mind, although I haven't checked whether that's the mechanism here), is there any obvious reason for discard this iteration over this bumping check?

**Comment by lynxlynxlynx** [comment]
Wouldn't exempting pcs just drop them on the next check? If that was not a problem it could easily lead to them walking through each other.

**Comment by MarcelHB** [comment]
That's just the path planning here, the actual can-bump-on-step test is in `DoStep`. If re-pathing is about to be done, this particular actor goes away anyway, unless there is a reason to deny `GA_ONLY_BUMPABLE` for actor. What came to my mind: Why repeat this whole calculation without considering actors blocking, if the reason that's preventing the first run to succeed is just the party itself, in the case of standing too close to a door?

Also, as I read the next check, it means it's blocked _unless_ it's passable, travellable (performable) or an actor (potentially bumpable).

**Comment by lynxlynxlynx** [comment]
Try it then. 🤷🏾

**Comment by MarcelHB** [comment]
Sounds good, doesn't work: Actually, the 2nd attempt fails as well in these cases.

**Comment by MarcelHB** [comment]
The bad news: When exhausted on a huge map, the algorithm is simply expensive, all the observed frame drops are a result of unfortunate geometry/searchmap blocking. Especially bad it is when the ignore-actors flags fails the second time for a whole group, because there is actually just no way, see BG2 slums:

![pathfinding](https://user-images.githubusercontent.com/238558/205415051-8b4d1e22-2583-4ac4-8edd-be3a7fcd3d09.gif)

Yet there is a way (I can submit a PR if interested) to reuse a little bit of information to halve the penalty by skipping the ignore-actors run if there wasn't any actor to potentially bump into.

The T* stuff doesn't hurt r/n on small maps but on larger ones, we get what we see if there is no path to find. Not sure what to do about this at the moment.

**Comment by lynxlynxlynx** [comment]
It's the worst case scenario, so not as bad as it may seem, but optimizations are always welcome.

I'm sceptical we could find a faster overall algorithm that would still satisfy all our needs. T* is already an optimization over the A*-like pathfinder we had before, which didn't even support all orientations, let alone bumping.

**Comment by bradallred** [comment]
I feel like we are possibly putting the cart before the horse. There are a ton of considerations to make when changing path finding.

I think we need to figure out exactly what is happening to know the best path forward (pun possibly intended).

It _seems_ like the main issue is that we completely recalculate paths at times when we could simply recalculate a portion. Is this not the case? I haven't thoroughly studied the code, but I don't recall seeing any attempts to modify existing paths.

Lynx also hinted that there are possibly issues when reaching the end of a path.

I think there are big improvements that could be made without having to do anything special.

**Comment by burner1024** [comment]
Maybe fixing #1785 would make a difference.

**Comment by MarcelHB** [comment]
This problem is even more extreme when playing on higher FPS.

I propose another idea: We apply some kind of flood algorithm in a way that two legally adjacent, non-blocking search map cells will be assigned the same number, as their neighbour, starting from top-left. If there is no such neighbour, they will be assigned a different number, so you end up with some kind of map of islands. This map is updated whenever there is a area load, or something mildly permanent changes (like a door). This is linear to the number of search map cells, executed rarely.

So if start end destination are on different search map islands, there is no need to initiate path-finding (save 6 times for a full party).

**Comment by lynxlynxlynx** [comment]
That wouldn't change anything for door approaches though. But it appears this issue is now hosting several issues and should be split into individual reports / discussions.

**Comment by lynxlynxlynx** [comment]
The title and first post are about "same island" problems, while the stuff about the sphere is not. With all the work that has already been done, the first part may be just a duplicate of #1785 now. Worst case scenarios like the separate slums area will require a different approach.

**Comment by MarcelHB** [comment]
Then, on the other hand, we could use this to distinguish the approaches. One case: skipping pathfinding completely, or otherweise trying something with actor bumping, moving, ordering instead, since we know it's a same-island problem.

**Comment by lynxlynxlynx** [comment]
It's just unfortunate that github doesn't allow splitting issues, you can only do one comment and then manually edit in all the rest.

**Comment by rsantellan** [comment]
Hello,
One thing that has changed is that now I have this on the console:
```
[PathFinderWIP/DEBUG]: Abandoning because I'm close to the goal
[WalkTo/DEBUG]: Aerie: Path was just abandoned
[PathFinderWIP/DEBUG]: Abandoning because I'm close to the goal
[WalkTo/DEBUG]: Aerie: Path was just abandoned
[GameScript/ERROR]: Invalid scripting action: 'runtopoint([1164.1059])'
[GameScript/ERROR]: Invalid scripting action: 'runtopoint([1164.1107])'
[GameScript/ERROR]: Invalid scripting action: 'runtopoint([1164.1011])'
[GameScript/ERROR]: Invalid scripting action: 'runtopoint([1100.1059])'
[GameScript/ERROR]: Invalid scripting action: 'runtopoint([1052.1059])'
[GameScript/ERROR]: Invalid scripting action: 'runtopoint([1004.1059])'
[KeyMap]: Looking up key: 137(�) 
[KeyMap]: Looking up key: 137(�) 
[KeyMap]: Looking up key: 137(�) 
```
The **[GameScript/ERROR]: Invalid scripting action: 'runtopoint([1052.1059])'** is new, didn't happen before, and I now I can't get out of places without a lot of work.

**Comment by lynxlynxlynx** [comment]
That's a pst-only action, other games don't support running. Harmless error, you j7st doubleclicked.

**Comment by lynxlynxlynx** [comment]
Opened a separate #2079 and closing this — the OP bug is fixed.
